move data files to the correct place on develop install

Zsailer opened the pull request:
This should address an issue we're seeing where data files are copied in a different location when installed using `pip install -e .` versus `pip install .`. It might be worth writing some more tests with different combinations of nested paths.

In short, this just strips the path from the data_files, using only their file names. This is consistent with [disutils' docs](https://docs.python.org/3.8/distutils/setupscript.html#installing-additional-files):
> No directory information from files is used to determine the final location of the installed file; only the name of the file is used. 

@blink1073

blink1073 commented:
We do need to test here that nested files work in `shared`, since that is how settings and theme files are handled in lab extensions.

blink1073 commented:
I think that means we have to generate new target paths for each sub path.

Zsailer commented:
This should be good to go. 

I've added a new fixture to create packages with different `data_file_spec`s. I've tested multiple nested data file paths.

I've also added a test of `pip install .` directly, which looked broken in jupyter/jupyter_server#258. It seems to be working okay.

We'll need to do some more investigating of the pyproject.toml issues, but that's a separate PR.

Zsailer commented on tests/test_develop.py:
```diff
@@ -1,19 +1,33 @@
 
-def test_develop(package_dir):
+data_files_specs = [
+    ("jupyter-packaging-test", "share", "test.txt"),
+    ("jupyter-packaging-test/level1", "share", "test.txt"),
+    ("jupyter-packaging-test", "level1/test", "test.txt"),
+    ("jupyter-packaging-test/level1/level2", "level1/test", "test.txt")
```
@blink1073, here are the different paths I've added for testing.

blink1073 commented on tests/test_develop.py:
```diff
@@ -1,19 +1,33 @@
 
-def test_develop(package_dir):
+data_files_specs = [
+    ("jupyter-packaging-test", "share", "test.txt"),
```
Ideally we would be able to specify this as `("jupyter-packaging-test", "share", "**.*")` I think

Zsailer replied:
Right. I guess I really need two things here... i) data_files_spec (represented as globs) and ii) the actual filepaths to create for the test.

I was using this one argument for both the globbed spec and the argument used to create these files on disk (which need to be more specific). I'll update

Zsailer replied:
Ok, this should be addressed in the latest commit.

blink1073 commented on tests/test_develop.py:
```diff
@@ -7,21 +7,29 @@
-    ("jupyter-packaging-test/level1", "share", "test.txt"),
-    ("jupyter-packaging-test", "level1/test", "test.txt"),
-    ("jupyter-packaging-test/level1/level2", "level1/test", "test.txt")
+    ("jupyter-packaging-test", "share", "*.*"),
```
I meant a full recursive check, since we won't know all of the sub folders

Zsailer replied:
Ah, I see. I'm running into issues here. This is where I see different behavior from `pip install -e .` versus `pip install .`. 

* `pip install -e .` keeps the subfolders from the recursive check in the final destination where the files are copied. 
* (I think) `pip install .` drop the subfolders from the final destination. 

I'm verifying now with some extra unit tests.

Zsailer commented on tests/test_datafiles_install.py:
```diff
@@ -0,0 +1,87 @@
+        "jupyter-packaging-test//level1/level2/test.txt"
+    ),
+    (
+        # data file source
+        ("level1/level2/test/test.txt",),
+        # data file spec
+        ("jupyter-packaging-test", "level1", "**/*"),
+        # data file target
+        "jupyter-packaging-test/level2/test/test.txt"
```
Ok, @blink1073, I think I've covered the recursive case here. Let me know what you think.

Zsailer commented:
I've refactored the data_files tests quite a bit. I've included a set of side-by-side tests that verify `pip install -e .` and `pip install .` work exactly the same. Everything looks good!

In the new tests, you can clearly see the source, data file specs, and expected target directory.
